kart/spatial_tree.py

```diff
@@ -396,125 +387,261 @@ def _find_geometry_column(fields):
+    DEFAULT_BITS_PER_VALUE = 20
+
+    def __init__(self, bits_per_value=None):
+        if bits_per_value is None:
+            bits_per_value = self.DEFAULT_BITS_PER_VALUE
+
+        assert bits_per_value % 2 == 0  # bits_per_value must be even.
+        self.BITS_PER_VALUE = bits_per_value
+        self.BITS_PER_ENVELOPE = 4 * self.BITS_PER_VALUE
+        self.BYTES_PER_ENVELOPE = self.BITS_PER_ENVELOPE // 8
+        self.VALUE_MAX_INT = 2 ** self.BITS_PER_VALUE - 1
+        self.ENVELOPE_MAX_INT = 2 ** self.BITS_PER_ENVELOPE - 1
+
+        self.BYTE_ORDER = "big"
+
+    def encode(self, envelope):
+        """
+        Encodes a (w, s, e, n) envelope where -180 <= w, e <= 180 and -90 <= s, n <= 90.
+        Scale each value to a unsigned integer of bitlength BITS_PER_VALUE such that 0 represents the min value (eg -180
+        for longitude) and 2**BITS_PER_VALUE - 1 represents the max value (eg 180 for longitude), then concatenates
+        the values together into a single unsigned integer of bitlength BITS_PER_VALUE, which is encoded to a byte array
+        of length BYTES_PER_ENVELOPE using a big-endian encoding.
+        """
+        integer = self._encode_value(envelope[0], -180, 180, math.floor)
+        integer <<= self.BITS_PER_VALUE
+        integer |= self._encode_value(envelope[1], -90, 90, math.floor)
+        integer <<= self.BITS_PER_VALUE
+        integer |= self._encode_value(envelope[2], -180, 180, math.ceil)
+        integer <<= self.BITS_PER_VALUE
+        integer |= self._encode_value(envelope[3], -90, 90, math.ceil)
+        assert 0 <= integer <= self.ENVELOPE_MAX_INT
+        return integer.to_bytes(self.BYTES_PER_ENVELOPE, self.BYTE_ORDER)
+
+    def _encode_value(self, value, min_value, max_value, round_fn):
+        assert min_value <= value <= max_value
+        normalised = (value - min_value) / (max_value - min_value)
+        encoded = round_fn(normalised * self.VALUE_MAX_INT)
+        assert 0 <= encoded <= self.VALUE_MAX_INT
+        return encoded
+
+    def decode(self, encoded):
+        """Inverse of encode_envelope."""
+        integer = int.from_bytes(encoded, self.BYTE_ORDER)
+        assert 0 <= integer <= self.ENVELOPE_MAX_INT
+        n = self._decode_value(integer & self.VALUE_MAX_INT, -90, 90)
+        integer >>= self.BITS_PER_VALUE
+        e = self._decode_value(integer & self.VALUE_MAX_INT, -180, 180)
+        integer >>= self.BITS_PER_VALUE
+        s = self._decode_value(integer & self.VALUE_MAX_INT, -90, 90)
+        integer >>= self.BITS_PER_VALUE
+        w = self._decode_value(integer & self.VALUE_MAX_INT, -180, 180)
+        return w, s, e, n
+
+    def _decode_value(self, encoded, min_value, max_value):
+        assert 0 <= encoded <= self.VALUE_MAX_INT
+        normalised = encoded / self.VALUE_MAX_INT
+        return normalised * (max_value - min_value) + min_value
+
+
+def get_envelope_for_indexing(geom, transforms, feature_oid):
```

Comment:
I went looking for actual GPKGs that cross the antimeridian and I found that they had values >180. I assumed that this was the proper way of encoding geometries in GPKG, since it is also the one the only one that makes sense with min-x and max-x envelopes that GPKG requires:
- if you have a polygon that has 4 different x values: [170, 175, 185, 190] then the min and max x values are 170 and 190, which is useful
- for the same geometry, if you store those x values as [170, 175, -175, -170] then the min and max x values are -170 and 170, which are useless - those values are not at the outside of the shape, and they don't tell us if the shape crosses the antimeridian or not, so there's nothing at all we can use them for.

However, you are right - just because it wouldn't make sense to encode a single geometry in GPKG which is split into two halves, one near 180 and the other near -180, because that breaks the min and max envelopes, doesn't mean that nobody is doing this.

I propose the following:
1. I construct an anticlockwise envelope around the min and max values of the geometry in its native CRS.
(This can mean I sometimes get a useless envelope for geometries that are split by the antimeridian - eg spanning from -179 to 179 - and sometimes I will get a useful envelope eg 179 to 181, depending on how the geometry is encoded).
2. I transform the envelope into EPSG:4326
3. I shift parts of the envelope by 360 degrees until the envelope has anticlockwise winding order.
(So far the algorithm is unchanged)
4. I test if the resulting envelope spans more than 180 degrees of longitude. If it does, I abandon the attempt to find a meaningful envelope and return None, so this geometry will match all spatial filters.

Why this should work:
The code I have is working for input geometries that don't cross the antimeridian, and that are contiguous across the antimeridian (eg spanning 179 to 181) - and the fact that these geometries are split in two halves post-transform I can handle fine - but my code won't work for input geometries that are split into two halves, since then (min-x, max-x) will be something like (-179, 179), which looks the same to me as a geometry that covers the whole globe except for a tiny gap near the antimeridian. However, since its width is 358 degrees, I just give up, so the end result will be fine.
The only case where this wouldn't work - where I won't detect that the shape has been split and simply give up - is if there is a single geometry that crosses the antimeridian with a single line segment which is > 180 degrees of longitude wide, but that doesn't seem like a real world geometry.

Why I want to do it this way:
This way keeps the efficiency of the current method - the rare case which is too hard to index using the current method, we just don't index. Inspecting underlying geometries to try and recalculate their true envelopes when they have been split across the antimeridian could require a) a deeper knowledge of the various CRSs than I currently have, and a lot more code for handling odd geometries and b) algorithms that are O(n^2) where n = num_points, running on the individual geometries with all of their complexity.

Reply:
Thinking a little bit more about this: the one case I said I can't handle correctly above is actually not a real-case since it looks identical from our point of view to a smaller geometry with shorter line segments that doesn't cross the antimeridian. The only way we could distinguish between the two would be to check the winding order of the raw geometry, but since we don't know what the winding order is supposed to be, we just have to assume the smaller geometry with smaller line segments is the right interpretation. So, I've convinced myself that this approach is adequate...

Reply:
We talked about this the other day but I've only just read it written down. What you've written is a good approach I think. WRT supporting large sparse-vertexed boxes that cover more than a hemisphere: as you've mentioned, there's no way to handle those without knowing the intended winding order in advance. Luckily, no one ever does that. But if they do, tough :)

Reply:
> I assumed that this was the proper way of encoding geometries in GPKG, since it is also the one the only one that makes sense with min-x and max-x envelopes that GPKG requires:

It's kind of up to the input geometry whether it goes past >180° or <-180° longitude... but geometries that use "normalised" coordinates certainly aren't wrong/invalid. There's no requirement in GPKG/anywhere else for extending beyond those limits. Since geometries are projected vertex-by-vertex, a projected CRS geometry transformed into 4326 will always have vertexes in the range (-180° <= x <= 180°).

Reply:
Seems good to me.

> The only case where this wouldn't work - where I won't detect that the shape has been split and simply give up - is if there is a single geometry that crosses the antimeridian with a single line segment which is > 180 degrees of longitude wide, but that doesn't seem like a real world geometry.

This is basically what MS SQL Server does, fwiw, and rejects any geometries with vertexes more than a hemisphere apart — your points need to be <=180° apart. I think it's a good approach.

Lots of tests to make sure that the behaviour works as documented :-)